arangod/RocksDBEngine/RocksDBEngine.cpp

```diff
@@ -1325,6 +1333,90 @@ RecoveryState RocksDBEngine::recoveryState() noexcept {
 TRI_voc_tick_t RocksDBEngine::recoveryTick() noexcept {
   return TRI_voc_tick_t(server().getFeature<RocksDBRecoveryManager>().recoveryTick());
 }
+void RocksDBEngine::compactRange(RocksDBKeyBounds bounds) {
+  {
+    WRITE_LOCKER(locker, _pendingCompactionsLock);
+    _pendingCompactions.push_back(std::move(bounds));
+  }
+
+  // directly kick off compactions if there is enough processing
+  // capacity
+  processCompactions();
+}
+
+void RocksDBEngine::processCompactions() {
+  Scheduler* scheduler = arangodb::SchedulerFeature::SCHEDULER;
+  if (scheduler == nullptr) {
+    return;
+  }
+
+  uint64_t maxIterations = _maxParallelCompactions;
+  uint64_t iterations = 0;
+  while (++iterations <= maxIterations) {
+    if (server().isStopping()) {
+      // don't fire off more compactions while we are shutting down
+      return;
+    }
+
+    RocksDBKeyBounds bounds = RocksDBKeyBounds::Empty();
+    {
+      WRITE_LOCKER(locker, _pendingCompactionsLock);
+      if (_pendingCompactions.empty() || 
+          _runningCompactions >= _maxParallelCompactions) {
+        // nothing to do, or too much to do
+        LOG_TOPIC("d5108", TRACE, Logger::ENGINES) 
+            << "checking compactions. pending: " << _pendingCompactions.size() 
+            << ", running: " << _runningCompactions; 
+        return;
+      }
+      // found something to do, now steal the item from the queue
+      bounds = std::move(_pendingCompactions.front());
+      _pendingCompactions.pop_front();
+      // set it to running already, so that concurrent callers of this method
+      // will not kick off additional jobs
+      ++_runningCompactions;
+    }
+    
+    LOG_TOPIC("6ea1b", TRACE, Logger::ENGINES) 
+          << "scheduling compaction for execution";
+    
+    bool queued = scheduler->queue(arangodb::RequestLane::CLIENT_SLOW, [this, bounds]() {
```

Comment:
Is it guaranteed that `this` lives long enough? In other words - can it happen that this lambda gets executed after the `RocksDBEngine` instance has been deleted?

Reply:
That is a good point :thinking:

Reply:
Addressed.